libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/transforms/ClasspathEntrySnapshotTransform.kt

```diff
@@ -6,26 +6,92 @@
 
 import org.gradle.api.artifacts.transform.*
+import org.gradle.api.file.DirectoryProperty
 import org.gradle.api.file.FileSystemLocation
+import org.gradle.api.provider.Property
 import org.gradle.api.provider.Provider
 import org.gradle.api.tasks.Classpath
+import org.gradle.api.tasks.Internal
+import org.jetbrains.kotlin.build.report.metrics.*
+import org.jetbrains.kotlin.gradle.report.BuildMetricsReporterService
+import org.jetbrains.kotlin.incremental.classpathDiff.ClassSnapshotGranularity
 import org.jetbrains.kotlin.incremental.classpathDiff.ClasspathEntrySnapshotExternalizer
 import org.jetbrains.kotlin.incremental.classpathDiff.ClasspathEntrySnapshotter
 import org.jetbrains.kotlin.incremental.storage.saveToFile
+import java.io.File
 
 /** Transform to create a snapshot of a classpath entry (directory or jar). */
 @CacheableTransform
-abstract class ClasspathEntrySnapshotTransform : TransformAction<TransformParameters.None> {
+abstract class ClasspathEntrySnapshotTransform : TransformAction<ClasspathEntrySnapshotTransform.Parameters> {
+
+    abstract class Parameters : TransformParameters {
+        @get:Internal
+        abstract val gradleUserHomeDir: DirectoryProperty
+
+        @get:Internal
+        abstract val buildMetricsReporterService: Property<BuildMetricsReporterService>
+    }
 
     @get:Classpath
     @get:InputArtifact
     abstract val inputArtifact: Provider<FileSystemLocation>
 
     override fun transform(outputs: TransformOutputs) {
-        val classpathEntry = inputArtifact.get().asFile
-        val snapshotFile = outputs.file(classpathEntry.name.replace('.', '_') + "-snapshot.bin")
+        val classpathEntryInputDirOrJar = inputArtifact.get().asFile
+        val snapshotOutputFile = outputs.file(classpathEntryInputDirOrJar.name.replace('.', '_') + "-snapshot.bin")
+
+        // There are two levels of granularity when taking a snapshot:
+        //   - CLASS_LEVEL (coarse-grained): The size of the snapshot is smaller, but we will have coarse-grained classpath changes, which
+        //     means more source files will be recompiled.
+        //   - CLASS_MEMBER_LEVEL (fine-grained): The size of the snapshot is larger, but we will have fine-grained classpath changes, which
+        //     means fewer source files will be recompiled.
+        // Therefore, CLASS_LEVEL is suitable for classes that are infrequently changed (e.g., external libraries which are typically
+        // stored/transformed inside the Gradle user home, plus a few hard-coded cases below), whereas CLASS_MEMBER_LEVEL is suitable for
+        // classes that are frequently changed (e.g., classes produced by the current project).
+        val granularity = if (
+            classpathEntryInputDirOrJar.startsWith(parameters.gradleUserHomeDir.get().asFile) ||
+            classpathEntryInputDirOrJar.name == "android.jar" ||
+            classpathEntryInputDirOrJar.name.startsWith("kotlin-compiler-embeddable")
```

Comment:
Please do not special-case for the Kotlin repo.

Reply:
Hm this (fat) jar is unusually large and it takes ~3s to transform it, so I wanted to special-case it, but sure I've removed it now and will deal with it later.